Improve gremlin-server.sh with start/stop support.

chupman opened the pull request:
Signed-off-by: Chris Hupman <chupman@us.ibm.com>
fixes an unfinished task on #845 

I combined the best of our current `gremlin-server.sh`, `janusgraph.sh` and Tinkerpop's `gremlin-server.sh` script to make a supergroup equivalent  of a service wrapper.

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there an issue associated with this PR? Is it referenced in the commit message?
- [X] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you written and/or updated unit tests to verify your changes?
- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [X] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [X] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

chupman commented:
@sjudeng and @twilmes For the unit file in this PR, and for file placement suggestions in general should we recommend that the `properties` and `server.yaml` files in use be stored in /etc/opt/? I want to be in line with what was decided on in janusgraph-docker.

pluradj commented:
@chupman do you have steps on how to test this?

chupman commented:
> @chupman do you have steps on how to test this? what platforms did you test against?

I've tested the script on mac and the unit file + script on ubuntu. I just added documentation as well. To test I would recommend:
1.  just use it like you would normally use gremlin-server.sh. Make sure you see the same behavior, it still defaults to `conf/gremlin-server/gremlin-server.yaml` for configuration.
2. Try `./bin/gremlin-server.sh (start|stop|restart|status)` ( again this uses `conf/gremlin-server/gremlin-server.yaml`)
3. If you have an System D or System V enabled os try the steps in the documentation and use either `service janusgraph (start|stop|restart|status)` or `systemctl (start|stop|restart|status) janusgraph` to verify

Since any issue or misleading documentation will heavily affects I'm really hoping I can get a couple people to try this out before merge. Lastly, I didn't test out install since it's something I've never used before and don't have a frame of reference.

chupman commented:
Realized I broke backwards compatibility and it currently only works with start|stop. Will fix later today and remove WIP.

chupman commented:
Updated default behavior to start the server in foreground using default file locations rather than showing the help message. Also tested based off my own recommendations. I believe this is in a good state now, but would still love for people to help test before a merge.

pluradj commented on docs/basics.adoc:
```diff
@@ -504,6 +504,51 @@ Connecting to Gremlin-Server (127.0.0.1:8182)... OK (connected to 127.0.0.1:8182
+[source, text]
+----
+ln -s bin/gremlin-server.sh /etc/init.d/janusgraph
+cp conf/janusgraph.env /etc/default/janusgraph
```
* symlink target should not use a relative path
* second line should use `ln -s`

chupman commented on janusgraph-dist/src/assembly/static/conf/janusgraph.service:
```diff
@@ -0,0 +1,21 @@
+WorkingDirectory=${JANUSGRAPH_HOME}
+# Using /bin/bash gets around potential SELinux issues with Java
+ExecStart=/bin/bash -c '${JANUSGRAPH_HOME}/bin/janusgraph-server.sh start'
+ExecStop=/bin/bash -c '${JANUSGRAPH_HOME}/bin/janusgraph-server.sh stop'
```
Updated to gremlin-server.sh. I was originally making a standalone script, but merged it with gremlin-server.sh

chupman commented on docs/basics.adoc:
```diff
@@ -504,6 +504,51 @@ Connecting to Gremlin-Server (127.0.0.1:8182)... OK (connected to 127.0.0.1:8182
+Symbolic Link
+[source, text]
+----
+ln -s conf/janusgraph.service /etc/systemd/system/janusgraph.service
```
I'm actually fine with totally removing the documentation. My real goal with these changes is to enable the creation of rpm and deb packages for JanusGraph, not to have people manually setup janusgraph as a service. I just need these changes in place before I can create packages.

Testing is really appreciated and I'll spin up some VMs to try things out again.

pluradj replied:
You shouldn't merge PR this unless it's clear to the reviewers on how to review.

If the manual testing steps are not intended for "public" consumption, you could put documentation in the GitHub issue.

If the `janusgraph.env` and `janusgraph.service` configs aren't intended for "public" consumption either, they should be in a different directory in the distribution other than `conf`.

chupman commented:
Personal TODO: figure out if I can facilitate adding meaningful logging in the event of java not being installed in journalctl. A useful message is already being logged at /var/log/janusgraph/janusgraph.log

chupman commented:
I just tested with centos 6 and 7 and everything is looking good. I'm going to try to retest ubuntu 14 and 18 by end of day.

chupman commented:
I only have an ubuntu 18 VM handy, but I tested both systemd and systemv installs so I feel like I have good coverage. I would some feedback on the new service wrapper.

chupman commented on janusgraph-dist/src/assembly/static/bin/gremlin-server.sh:
```diff
@@ -1,45 +1,118 @@
-CP="$CP":$(find -L $EXT -name '*.jar' | sort | tr '\n' ':')
+  done
+  JANUSGRAPH_BIN="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+  echo"Hello"
```
Well that's embarrassing. Thanks for pointing that out.

chupman commented on janusgraph-dist/src/assembly/static/bin/gremlin-server.sh:
```diff
@@ -1,45 +1,118 @@
+  # Add the remaining jars in $JANUSGRAPH_HOME/lib.
+  JANUSGRAPH_CP="$JANUSGRAPH_CP":$(find -L $JANUSGRAPH_LIB -name '*.jar' \
+                  \! -name 'janusgraph*' \
+                  \! -name 'slf4j-log4j12*.jar' | sort | tr '\n' ':')
```
It's actually saying to not add those last 2 lines because of the `!`. I double checked in the man page to be safe.
```
find / \! -name "*.c" -print
             Print out a list of all the files whose names do not end in .c.
```

chupman commented on janusgraph-dist/src/assembly/static/bin/gremlin-server.sh:
```diff
@@ -50,18 +123,160 @@ fi
+    exit 1
+  fi
+
+  if [[ -z "$RUNAS" ]]; then
```
The TinkerPop upstream has RUNAS for `start` and `install`. Looks like I just missed them when I was merging the various files. Thanks for the catch.

chupman replied:
I added RUNAS for start and install. I did some initial testing on my Mac, but will want to do a little more before merging probably. 

I found out today that  `su -c` is different in osx so I had to add OS detection.

chupman commented on janusgraph-dist/src/assembly/static/conf/janusgraph.service:
```diff
@@ -0,0 +1,16 @@
+[Service]
+EnvironmentFile=/opt/janusgraph/conf/janusgraph.env
+Type=forking
+
```
I think it seems like a good idea. Added.

chupman commented on docs/basics.adoc:
```diff
@@ -504,6 +504,38 @@ Connecting to Gremlin-Server (127.0.0.1:8182)... OK (connected to 127.0.0.1:8182
 
+==== Running JanusGraph Server as a daemon
+
+The JanusGraph https://github.com/JanusGraph/janusgraph/releases[release] also comes with an init script that can be found at `bin/gremlin-server.sh`. The `gremlin-server.sh` script was upgraded to an init script which supports start, stop, status. Its usage is backwards compatible and can still be run in the foreground by invoking `./bin/gremlin-server.sh` with an option parameter of a gremlin s ...
```
Thanks for the link. Since markdown GitHub has the same behavior of not displaying the line breaks I'll try to follow this going forward. Looks like I learned something else today.

pluradj commented on docs/basics.adoc:
```diff
@@ -504,6 +504,39 @@ Connecting to Gremlin-Server (127.0.0.1:8182)... OK (connected to 127.0.0.1:8182
+==== Running JanusGraph Server as a Daemon
+
+The JanusGraph https://github.com/JanusGraph/janusgraph/releases[release] also comes with an init script that can be found at `bin/gremlin-server.sh`.
+The script can also be run in the foreground by invoking `./bin/gremlin-server.sh` with an optional parameter of a Gremlin Server yaml file.
```
Running JanusGraph as a daemon as described here only starts the Gremlin Server. The default pre-packaged distribution quickstart runs `janusgraph.sh` which starts Cassandra and Elasticsearch, so there's a bit of a disconnect from the previous section that should be documented here. When running JanusGraph as a daemon, it is expected that the operator will install/run the storage and indexing backends separately.

FlorianHockmann replied:
I wonder whether it would make sense to convert `janusgraph.sh` into an init script that can be run as a service. I think that would give users the easiest way to get started with JanusGraph.
That could of course be done in a different PR. If we want to do that, then we could document here how `janusgraph.sh` can be used as a service and move the documentation about only running Gremlin Server into the JanusGraph Server chapter.

pluradj replied:
The pre-packaged distribution discussion probably needs to move earlier anyway, like Chapter 3 Getting Started.

FlorianHockmann replied:
Oh yes, I just assumed that this was part of the introduction because of the filename being `basics.adoc`, but then I agree that we should move in in a separate PR.

chupman replied:
I don't think it would take much for `janusgraph.sh` to work as an init script since it already supports `start/start/clean/status`.

Separately I had a different idea for easiest way to get started from a [medium article](https://medium.com/@patternrecognizer/how-to-add-a-try-in-play-with-docker-button-to-your-github-project-41cb65721e94) I read awhile back. I was just waiting for janusgraph-docker, and now the 0.2.3 release, to finish up.

I opened up issues for moving the `janusgraph.sh` docs, testing it as a init script, and creating a play with docker button. Still need to incorporate Jason's feedback about specifying that it's just running Gremlin server.

FlorianHockmann replied:
Great idea with the play with docker button, I just want to node that I see `janusgraph.sh` more as something that can be used to actually set up a small JanusGraph database. We even used that in production in the beginning like that with a 3 node cluster.

chupman commented on janusgraph-dist/src/assembly/static/conf/janusgraph.env:
```diff
@@ -0,0 +1,4 @@
+JANUSGRAPH_YAML=/opt/janusgraph/conf/gremlin-server/gremlin-server.yaml
```
For the rpm and deb implementations I was planning to use `/etc/opt/janusgraph` to match linux standards and the docker image. The problem is that the server yamls have relative paths so changing the directories for these files will make things a little harder on users if they choose to manually setup.

In `janusgraph-docker` there is just one properties file named `janusgraph.properties`. I think it would be cleaner to do the same and just document that users should copy over whichever template they want to use. It might work best to make copies of the properties and yaml files with updated paths that just get used by rpm and deb packages. It's really import to get this detail right and I'll give this some more thought. 

I'm thinking of a structure like this right now

```
/etc/opt/janusgraph/janusgraph.properties
/etc/opt/janusgraph/janusgraph-server.yaml
/etc/opt/janusgraph-script.groovy
/etc/opt/janusgraph/conf/*.yaml
/etc/opt/janusgraph/conf/*.properties
```
From with the `/etc/opt/janusgraph` directory users could change templates like `cp conf/janusgraph-cassandra-es.properties janusgraph.properties`

chupman replied:
A point I failed to address was using `JANUSGRAPH_CFG` instead would be much better for this. I still need to test out how to fix the issues of the paths within the gremlin-server.yaml file.

pluradj replied:
Why `/etc/opt/janusgraph` instead of `/etc/janusgraph`?

Normalizing and simplifying the file names in the config directory would be good.

* graph.properties
* server.yaml
* server.groovy

Of course, this is assuming `JanusGraphFactory` as opposed to `ConfiguredGraphFactory`, in which case the script file might not be as necessary with the dynamic graph bindings.

chupman replied:
I asked the same thing in the initial janusgraph-docker pr and got schooled in this [comment](https://github.com/JanusGraph/janusgraph-docker/pull/1#pullrequestreview-151053377)

Here's the link to the [spec from the linuxfoundation](http://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s07.html#etcoptConfigurationFilesForOpt)
My main concern is to be consistent across the board.

chupman replied:
@pluradj I just pushed an additional commit with `/etc/opt/janusgraph` as the `JANUSGRAPH_CFG` default. Since `gremlin-server.yaml` doesn't care about `JANUSGRAPH_CFG` I had to hardcode the new paths for the properties and script files in a new template file, `daemon-gremlin-server.yaml`

pluradj commented:
@chupman and I were discussing this today, and the daemon support seems to be a new feature separate from the `gremlin-server.sh` updates. The `gremlin-server.sh` should stick here in `0.2` since the updates made are aligned with the TinkerPop version. I'd suggest moving the daemon support out into a separate PR and targeting it for `0.4`.

chupman commented:
@FlorianHockmann @pluradj I updated the PR to just include the updated gremlin-server.sh script. I believe all the open comments have already been address. Can you please do another review?

I'll create separate pull requests targeting 0.4 with the config files I removed.
